common/datasets/sms_wsj/returnn_datasets.py

```diff
@@ -0,0 +1,617 @@
+            hdf_classes[speaker_idx] = np.concatenate(
+                [
+                    self._pad_label * np.ones(pad_start),
+                    hdf_classes[speaker_idx],
+                    self._pad_label * np.ones(pad_end),
+                ]
+            )
+        d["target_classes"] = np.stack(hdf_classes).T
+        d["target_classes_len"] = np.array(padded_len)
+        return d
+
+    def get_seq_length_for_keys(self, seq_idx: int) -> NumbersDict:
+        """
+        Returns sequence length for all data/target keys.
+        """
+        d = super().get_seq_length_for_keys(seq_idx)
+        data = self[seq_idx]
+        d["target_classes"] = int(data["target_classes_len"])
+        return NumbersDict(d)
+
+
+class SmsWsjWrapper(MapDatasetWrapper):
+    """
+    Base class for datasets that can be used in RETURNN config.
+    """
+
+    def __init__(self, sms_wsj_base, **kwargs):
+        """
+        :param Optional[SmsWsjBase] sms_wsj_base: SMS-WSJ base class to allow inherited classes to modify this
+        """
+        if "seq_ordering" not in kwargs:
+            print("Warning: no shuffling is enabled by default", file=returnn_log.v2)
+        super().__init__(sms_wsj_base, **kwargs)
+        # self.num_outputs = ...  # needs to be set in derived classes
+
+        def _get_seq_length(seq_idx: int) -> NumbersDict:
+            """
+            Returns sequence length for all data/target keys.
+            """
+            corpus_seq_idx = self.get_corpus_seq_idx(seq_idx)
+            return sms_wsj_base.get_seq_length_for_keys(corpus_seq_idx)
+
+        self.get_seq_length = _get_seq_length
+
+    @staticmethod
+    def _pre_batch_transform(inputs: Dict[str, Any]) -> Dict[str, np.array]:
+        """
+        Used to process raw SMS-WSJ data
+        :param  inputs: input as coming from SMS-WSJ
+        """
+        return_dict = {
+            "seq_tag": np.array(inputs["example_id"], dtype=object),
+            "source_id": np.array(inputs["source_id"], dtype=object),
+            "seq_len": np.array(inputs["num_samples"]["observation"]),
+        }
+        return return_dict
+
+    def _collect_single_seq(self, seq_idx: int) -> DatasetSeq:
+        """
+        :param seq_idx: sorted seq idx
+        """
+        corpus_seq_idx = self.get_corpus_seq_idx(seq_idx)
+        self._dataset.update_buffer(corpus_seq_idx)
+        data = self._dataset[corpus_seq_idx]
+        assert "seq_tag" in data
+        return DatasetSeq(seq_idx, features=data, seq_tag=data["seq_tag"])
+
+    def init_seq_order(self, epoch: Optional[int] = None, **kwargs) -> bool:
+        """
+        Override this in order to update the buffer. get_seq_length is often called before _collect_single_seq,
+        therefore the buffer does not contain the initial indices when continuing the training from an epoch > 0.
+        """
+        out = super().init_seq_order(epoch=epoch, **kwargs)
```

Comment:
```suggestion
        self.seq_ordering = self._seq_ordering
        out = super().init_seq_order(epoch=epoch, **kwargs)
```

Not sure if this is relevant in general, but when using `ReturnnSearchJobV2`, `self.seq_ordering` might be changed by `def search()` in `returnn/tf/engine.py`, which leads to problems with the `update_buffer` function, so it has to be reset

Reply:
Could you please elaborate what the issue is in `update_buffer`?

Reply:
Ok, I'm sorry, what I said was not 100% right. I recreated the problem. The issue is not `update_buffer`.
In my case, `self.seq_ordering` was changed to `"sorted_reverse"`. The `init_seq_order()` function in `MapDatasetWrapper` calls `self.get_seq_order_for_epoch`. If `self.seq_ordering` is `"sorted_reverse"`, `get_seq_len()` is called and it tries to get the sequence length for sequences which are not in the buffer. Therefore, `self.seq_ordering` has to be `"default"`, so not all sequence lengths are needed. Maybe there is a better and more elegant solution to this problem, but this was the easiest one I could think of